compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3981,7 +3981,9 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
 
     /** Convert constructor proxy reference to a new expression */
     def newExpr =
-      val Select(qual, nme.apply) = tree: @unchecked
+      val qual = (tree: @unchecked) match
+        case Select(qual, nme.apply) => qual
+        case Ident(nme.apply) => This(tree.symbol.owner.asClass).withSpan(tree.span)
```

Comment:
I think this doesn't handle unqualifed `apply` that are resolved by an import:
```scala
object X { class C(); object C; object D { import C.*, apply() } }
```
More robust:
```suggestion
        case tree @ Ident(nme.apply) => tpd.desugarIdentPrefix(tree)
```
```